File: 20_valid_parentheses.js

//https://leetcode.com/problems/valid-parentheses/

/**
 * @param {string} s
 * @return {boolean}
 *
 * Accepted
 */

var isValid = function(s) {
    if (s === '') {
        return true;
    }
    if (s.length % 2 !== 0) {
        return false;
    }
    //使用栈来实现
    var stack = [];

    var i = 0,
        l = s.length;

    for (; i < l; i ++) {
        var c = s[i];
        if (c === '(' || c === '[' || c === '{') {
            stack.push(c);
            continue;
        }
        var top = stack.pop();
        if (c === ')' && top !== '(') {
            return false;
        }
        if (c === ']' && top !== '[') {
            return false;
        }
        if (c === '}' && top !== '{') {
            return false;
        }
    }
    //如果这个时候，栈不是空的话，return false
    if (stack.length) {
        return false;
    }
    return true;
};

File: python/20_valid_parentheses.py
#!/usr/bin/env python3
# coding: utf-8

# https://leetcode.com/problems/valid-parentheses/

class Solution:
    # @param {string} s
    # @return {boolean}
    # Accepted
    def isValid(self, s):
        l = len(s)
        if l == 0:
            return True

        if l % 2 != 0:
            return False

        stack = []

        for c in s:
            #print(c)
            if c == '(' or c == '[' or c == '{':
                #入栈
                stack.append(c)
                continue
            #javascript不会报错
            if len(stack) == 0:
                return False

            end = stack.pop()

            if c == ')' and end != '(':
                return False
            if c == ']' and end != '[':
                return False
            if c == '}' and end != '{':
                return False

        if len(stack) != 0:
            return False

        return True
